### PR TITLE
feat(functions): Mode aware functions

### PR DESCRIPTION
**Pull request by hinell**
https://github.com/mrjones2014/legendary.nvim/assets/8136158/83c5a565-228e-46fd-9059-fcc43d1758b6

**Summary:**
- feat(functions): populate `mode_mappings`
- feat(functions): add methods for UI display filtering
- feat(functions): allow filtering by functions
- feat(functions): display multi-mode functions
- docs(functions): describe new mode=field

> [!NOTE]
> I request to squash-rebase

## How to Test

New `funcs = { ... mode = 'n' }` field for functions - its type should be either `string` or `table<string>`; that's it

## Testing for Regressions

I have tested the following:

- [x] Triggering keymaps from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating keymaps via `legendary.nvim`, then triggering via the keymap in all modes (normal, insert, visual)
- [x] Triggering commands from `legendary.nvim` in all modes (normal, insert, visual)
- [x] Creating commands via `legendary.nvim`, then running the command manually from the command line
- [x] `augroup`/`autocmd`s created through `legendary.nvim` work correctly

### CODE REVIEW

File: doc/table_structures/FUNCTIONS.md
Comment (mrjones2014) [inline doc/table_structures/FUNCTIONS.md:38]:
```suggestion
You can use `mode` property to narrow function's scope, so it only appears in the specified mode(s):
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:37]:
```suggestion
      'item.mode must be a string or list of strings specifying modes: n, i, v etc.',
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:67]:
```suggestion
      local current_mode = vim.fn.mode()
      if not vim.tbl_contains(instance.mode_mappings, current_mode) then
        return
      end
      impl()
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:72]: It doesn't feel necessary to me to cache this.

```suggestion
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:94]: Based on the diff it looks like this function isn't needed either.

```suggestion
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:95]:
```suggestion
---@return string[] | nil
function Function:modes()
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:102]: let's keep it consistent with how the `mode_mappings` API works for other items.

```suggestion
  if vim.tbl_islist(self.mode_mappings) then
    return self.mode_mappings
  end

  local modes = {}
  for mode, mapping in pairs(self.mode_mappings) do
    if mapping then
      table.insert(modes, mode)
    end
  end

  if #modes == 0 then
    return { 'n', 'V', 'v', 'x', 's', 'o', '' }
  end

  return modes
```

File: lua/legendary/ui/format.lua
Comment (mrjones2014) [inline lua/legendary/ui/format.lua:41]: Why is this here?

File: lua/legendary/ui/format.lua
Comment (mrjones2014) [inline lua/legendary/ui/format.lua:49]: We shouldn't be ignoring formatting unless there's a good reason.

File: lua/legendary/ui/format.lua
Comment (mrjones2014) [inline lua/legendary/ui/format.lua:45]:
```suggestion
      table.concat(item--[[@as Function]]:modes() or { Config.icons.fn }, ', '),
```

File: lua/legendary/data/function.lua
Comment (mrjones2014) [inline lua/legendary/data/function.lua:22]: We shouldn't have the `mode_mappings` be a different type for functions than for everything else.

Take a look at how `Keymap` handles it. The type should be something similar to `ModeKeymap`, but you'll just need to modify it slightly to take different options (the options for a `Function`).

**Review by mrjones2014 — CHANGES_REQUESTED** [review]

**Comment by mrjones2014** [comment]
Also, please update the PR title per [CONTRIBUTING.md](https://github.com/mrjones2014/legendary.nvim/blob/master/CONTRIBUTING.md#prs)

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:22]: Quite a lot for an overnight PR which I made without a prior knowledge of the project, isn't it?

I didn't intend to implement fully `keymap`-compatible mode-map, cuz:
* Nobody is going to drop funcs items into `legendary.keymap({ ... })`
* Even if they do, current modemap type is fully forward compatible with `legendary.keymap({ ... })`;
* Zero use cases



> _you need just slightly_

Only to make things more complicated: validation (see below) requires that item's `[1]`-field is provided with a function, so introduction of `keymaps`-like node-map will likely break something  - you have to figure you what to do with that function; it's subject of further discussion.

I suggest you take care of map enhancement further after merging this PR.

https://github.com/mrjones2014/legendary.nvim/blob/6ea46c285f3fd49317e04b6b31fad7c86b975cb0/lua/legendary/data/function.lua#L15-L20

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:94]: It's used in UI display

File: lua/legendary/ui/format.lua
Comment (hinell) [reply lua/legendary/ui/format.lua:45]: Please, look carefully what you suggest; this introduces a bug; `item:modeSwitched()` is essential here.

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:72]: It's used for displaying: https://github.com/mrjones2014/legendary.nvim/pull/415/commits/29513d11cf20d125ad9a26645ab7d31b43db39da#r1366849251 

; It's essential for both performance reason and readability

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:94]: It's used for displaying: https://github.com/mrjones2014/legendary.nvim/pull/415/commits/29513d11cf20d125ad9a26645ab7d31b43db39da#r1366849251

File: lua/legendary/data/function.lua
Comment (mrjones2014) [reply lua/legendary/data/function.lua:22]: > Quite a lot for an overnight PR which I made without a prior knowledge of the project, isn't it?

Not sure what you mean. There's no deadline on the PR.

Regarding your 3 bullets, addressing in-order:

- That's not what I said and it's not what my comment is meant to support
- See above
- The use case is to support the alternate declaration syntax (2nd code block [here](https://github.com/mrjones2014/legendary.nvim/blob/master/doc/table_structures/KEYMAPS.md#specifying-mode))

> Only to make things more complicated: validation (see below) requires that item's [1]-field is provided with a function, so introduction of keymaps-like node-map will likely break something - you have to figure you what to do with that function; it's subject of further discussion.

It's not really more complicated. You can copy the validator from `keymap.lua` only for the options, validate that they're `Function` options instead of `Keymap` options. Then just update `executor.lua` to pull the right implementation out of the item based on the mode.

It we're going to introduce `mode_mappings` for `Function`s, then they're going to work the same as they do for `Keymap`s.

File: lua/legendary/data/function.lua
Comment (mrjones2014) [reply lua/legendary/data/function.lua:102]: Disagree.

File: lua/legendary/data/function.lua
Comment (mrjones2014) [reply lua/legendary/data/function.lua:94]: Yeah, you don't have to do it that way though -- see the other comments.

File: lua/legendary/ui/format.lua
Comment (hinell) [reply lua/legendary/ui/format.lua:41]: Stylua makes lines in-between ugly otherwise

File: lua/legendary/ui/format.lua
Comment (mrjones2014) [reply lua/legendary/ui/format.lua:45]: It doesn't introduce a bug if you also tweak the implementation of `modes()` as I suggested.

File: lua/legendary/ui/format.lua
Comment (mrjones2014) [reply lua/legendary/ui/format.lua:41]: I'd rather just have the formatter format it than stray from the style of everything else in the entire codebase.

File: lua/legendary/data/function.lua
Comment (mrjones2014) [reply lua/legendary/data/function.lua:72]: I really don't think it will have a noticeable impact on performance, and I don't see how it affects readability either, personally.

If you can show some benchmarks proving otherwise I'm all ears.

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:22]: > (2nd code block [here](https://github.com/mrjones2014/legendary.nvim/blob/master/doc/table_structures/KEYMAPS.md#specifying-mode))
>  then they're going to work the same as they do for Keymaps.

As I said, this PR is _about_ functions, not keymaps; if you want to unify mode-map apis for both, I would suggest you to take care of further mode-map improvement yourself taking the laid groundwork as basis; for me, it's more than enough to have `mode = { 'n', 'v', 'x' }` per one function; for more - i would just create more items with different `mode`; cheap and quick

And as I said, I see no reason to introduce that complicated approach, cause there are going to be [ambiguities](https://github.com/mrjones2014/legendary.nvim/pull/415#discussion_r1367045802) you will have to resolve as project owner

**Comment by hinell** [comment]
If you want more features - go ahead, I'm fine; 

My point is: I suggest the laid ground work for futher improvement, it's obviously beneficial to have half of the job done, then doing it yourself

It's also faster to merge and improve, than bogging down this discussion in very vague, or obviously incorrect suggestions

It's sad that I have to maintain a separate fork of this project; anyway

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:102]: Not an intent of this PR; I'm fine if you refactor it further after merging it;

there is a lof of space for improvement in this project

**Comment by mrjones2014** [comment]
> it's obviously beneficial to have half of the job done

Sure, you're definitely right about that. That doesn't mean I'm going to merge the feature in its current half-baked state.

> It's also faster to merge and improve

Yes it's faster, that doesn't mean I'm okay with merging half-baked features that don't align with existing features for other item types.

> than bogging down this discussion in very vague, or obviously incorrect suggestions

Please explain to me how you, who by your own admission doesn't have any prior knowledge of the project, would be able to identify that better than the project maintainer, and point me to a suggestion I made that is "obviously incorrect".

> It's sad that I have to maintain a separate fork of this project; anyway

I'm more than willing to work with you to get the PR into a mergeable state, however you don't seem willing to accept any of the feedback or address any of the problems with the PR.

If you'd rather maintain your own fork than actually collaborate, then go ahead, but open source collaboration doesn't mean merging incomplete PRs just because you say so.

Maybe you could try being a little less rude next time 🙂

File: lua/legendary/data/function.lua
Comment (hinell) [reply lua/legendary/data/function.lua:72]: @mrjones2014 It's cheap to check once and leave, then checking whether the list is empty all the time and it doesn't check. This is obvious.

Comment (mrjones2014) [reply lua/legendary/data/function.lua:72]: It’s an extremely inexpensive check. It doesn’t need to be cached.

**Comment by hinell** [comment]
> would be able to identify that better than the project maintainer,

As a maintainer, any kind of unity or 'consistency' you want across API's (`funcs()`, `keymaps(...)` etc.) - I'm fine - it's is ENTIRELY up to you to decide & implement; go ahead, write docs, make new feature - I'm fine; I suggest better - unite effort, merge this PR and improve it; there is no deadline on new features that aren't documented or demand strict "consistency" (What is this even?); nothing is goint to be broken, no harm, cheap and quick.

> Maybe you could try being a little less rude next time 🙂

Above you insisted on implementing entire keymap's modemap API for funcs (that's quite rude) borrowing from remotely related `keymaps` API, when my PR didn't intend to do that obviously.  What did you expect? :thinking: This is not collaborative in any way; that's serious extra work you want to be done; why asking at all? You shouldn't really

> its current half-baked state

Alright, when the half of the job is already done and you are saying the other half is going to be super easy-pisy to finish off, I suggest you go ahead and merge this and have what you have. There are no deadlines on refactoring. My best.

**Comment by mrjones2014** [comment]
I don't like to make a habit of merging incomplete features 🙂

**Comment by hinell** [comment]
Try to be less rude next time. Contribs & PRs come to you for free; we waste spare time and headache, not you, be respectful, don't ask for more work than necessary. There are no specs on what is complete and what is not. Grow up.

**Comment by hinell** [comment]
Try to be less rude next time. Contribs & PRs come to you for free; we waste spare time and headache, not you, be respectful, don't ask for more work than necessary. There are no specs on what is complete and what is not. Grow up.

**Comment by mrjones2014** [comment]
Deleted a bunch of useless vile rudeness comments.

**Pull request closed without merging**
